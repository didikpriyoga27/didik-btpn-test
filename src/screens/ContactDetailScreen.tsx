import {
  NavigationProp,
  RouteProp,
  useNavigation,
  useRoute,
} from '@react-navigation/native';
import {useMutation, useQuery, useQueryClient} from '@tanstack/react-query';
import axios from 'axios';
import React, {useEffect, useMemo, useState} from 'react';
import {
  ActivityIndicator,
  ScrollView,
  TextInput,
  ToastAndroid,
  TouchableOpacity,
} from 'react-native';
import FastImage from 'react-native-fast-image';
import {StackParamList} from '../navigation/StackParamList';
import Text from '../shared/components/Text';
import View from '../shared/components/View';
import {baseUrl} from '../slices/contactSlice';
import Header from '../shared/components/Header';

type ContactDetailScreenRouteProp = RouteProp<
  StackParamList,
  'ContactDetailScreen'
>;

export default function ContactDetailScreen() {
  const queryClient = useQueryClient();

  const {params} = useRoute<ContactDetailScreenRouteProp>();
  const {goBack} = useNavigation<NavigationProp<StackParamList>>();

  const [firstName, setFirstName] = useState('');
  const [lastName, setLastName] = useState('');
  const [age, setAge] = useState('');
  const [photo, setPhoto] = useState('');

  const defaultAvatar = useMemo(() => {
    if (params.contactId) {
      return null;
    }
    return `https://ui-avatars.com/api/?name=${firstName}+${lastName}&background=0f766e&color=fff`;
  }, [firstName, lastName, params.contactId]);

  const {data: contact, isLoading} = useQuery({
    queryKey: ['contact', params.contactId],
    queryFn: () => {
      if (params.contactId) {
        return axios
          .get(`${baseUrl}/${params.contactId}`)
          .then(result => result.data.data);
      }
      return null;
    },
  });

  useEffect(() => {
    if (contact) {
      setFirstName(contact?.firstName);
      setLastName(contact?.lastName);
      setAge(contact?.age?.toString());
      setPhoto(contact?.photo);
    }
  }, [contact]);

  const {mutateAsync: mutateSaveContact, isPending: isSaveLoading} =
    useMutation({
      mutationFn: async () => {
        axios.post(`${baseUrl}`, {
          firstName,
          lastName,
          age: Number(age),
          photo: defaultAvatar,
        });
      },
    });

  const {mutateAsync: mutateEditContact, isPending: isEditLoading} =
    useMutation({
      mutationFn: async () => {
        if (params.contactId) {
          const response = axios.put(`${baseUrl}/${params.contactId}`, {
            firstName,
            lastName,
            age: Number(age),
            photo,
          });
          return response;
        }
      },
    });

  const handleSave = async () => {
    if (params.contactId) {
      return mutateEditContact().then(() => {
        ToastAndroid.show('Contact edited successfully', ToastAndroid.SHORT);
        queryClient.refetchQueries({queryKey: ['contacts']});
        queryClient.refetchQueries({queryKey: ['contacts', params.contactId]});
        setTimeout(() => {
          goBack();
        }, 500);
      });
    }
    return mutateSaveContact().then(() => {
      ToastAndroid.show('Contact saved successfully', ToastAndroid.SHORT);
      queryClient.refetchQueries({queryKey: ['contacts']});
      setTimeout(() => {
        goBack();
      }, 500);
    });
  };

  if (isLoading) {
    return (
      <View className="flex-1 items-center justify-center">
        <ActivityIndicator size="large" />
      </View>
    );
  }

  return (
    <View className="flex-1">
      <Header title={'Contact Detail'} isWithBackButton />
      <ScrollView keyboardShouldPersistTaps={'handled'} className="flex-1">
        <FastImage
          source={{uri: params.contactId ? photo : defaultAvatar ?? ''}}
          className="w-24 h-24 mr-4 rounded-full bg-gray-300 self-center mt-4"
        />
        <Text className="px-4 mt-4 font-bold">First Name:</Text>
        <TextInput
          className="border border-gray-300 px-4 mx-4 rounded-md text-black dark:text-white"
          value={firstName}
          onChangeText={setFirstName}
          placeholder="Enter First Name"
        />
        <Text className="px-4 mt-4 font-bold">Last Name:</Text>
        <TextInput
          className="border border-gray-300 px-4 mx-4 rounded-md text-black dark:text-white"
          value={lastName}
          onChangeText={setLastName}
          placeholder="Enter Last Name"
        />
        <Text className="px-4 mt-4 font-bold">Age:</Text>
        <TextInput
          className="border border-gray-300 px-4 mx-4 rounded-md text-black dark:text-white"
          value={age}
          onChangeText={setAge}
          keyboardType="numeric"
          placeholder="Enter Age"
        />
      </ScrollView>
      <TouchableOpacity
        className="bg-teal-700 m-4 rounded-md h-12 justify-center"
        disabled={!firstName || !lastName || !age}
        onPress={handleSave}>
        <View className={'bg-transparent'}>
          {isEditLoading || isSaveLoading ? (
            <ActivityIndicator color={'white'} size={'small'} />
          ) : (
            <Text className="text-white text-center font-bold text-lg">
              Save
            </Text>
          )}
        </View>
      </TouchableOpacity>
    </View>
  );
}
