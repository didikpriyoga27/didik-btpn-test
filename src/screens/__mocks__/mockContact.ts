import {Contact} from '../../slices/contactSlice';

export const mockContact: Contact = {
  age: 20,
  firstName: 'John',
  lastName: 'Doe',
  id: 'id-1',
  photo:
    'https://ui-avatars.com/api/?name=John+Doe&background=0f766e&color=fff',
};

export const mockContact2: Contact = {
  ...mockContact,
  firstName: 'Ahmad',
  id: 'id-2',
};

export const mockContact3: Contact = {
  ...mockContact,
  firstName: 'Zayn',
  id: 'id-3',
};

export const mockContact4: Contact = {
  ...mockContact,
  firstName: 'John',
  id: 'id-4',
};
