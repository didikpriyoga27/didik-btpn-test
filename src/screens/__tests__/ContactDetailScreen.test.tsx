import {fireEvent, render} from '@testing-library/react-native';
import React from 'react';

import ContactDetailScreen from '../ContactDetailScreen';
import {mockContact} from '../__mocks__/mockContact';

jest.useRealTimers();

jest.mock('@react-navigation/native', () => {
  return {
    useNavigation: () => ({
      goBack: jest.fn(),
    }),
    useRoute: () => ({
      params: {
        contactId: 'id-1',
      },
    }),
  };
});

jest.mock('@tanstack/react-query', () => ({
  useQueryClient: jest.fn(() => ({
    refetchQueries: jest.fn(),
  })),
  useQuery: jest.fn(() => ({
    data: mockContact,
  })),
  useMutation: jest.fn(() => ({
    mutateAsync: jest.fn(() => Promise.resolve({})),
  })),
}));

describe('@screens/ContactDetailScreen', () => {
  const component = () => render(<ContactDetailScreen />);

  test('snapshot: component', async () => {
    const {toJSON, getByTestId} = component();
    const ButtonSave = getByTestId('ButtonSave');
    fireEvent(ButtonSave, 'press');
    expect(toJSON()).toMatchSnapshot();
  });
});
