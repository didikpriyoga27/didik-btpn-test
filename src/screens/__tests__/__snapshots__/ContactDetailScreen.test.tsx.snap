// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@screens/ContactDetailScreen snapshot: component 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#fff",
      },
      [
        {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ],
    ]
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "#fff",
        },
        [
          {
            "zIndex": 10,
          },
          {
            "width": "100%",
          },
          {
            "justifyContent": "flex-end",
          },
          {
            "elevation": 4,
            "height": 48,
            "marginTop": -0,
          },
        ],
      ]
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "#fff",
          },
          [
            {
              "flexDirection": "row",
            },
            {
              "alignItems": "center",
            },
            {
              "paddingLeft": 24,
              "paddingRight": 24,
            },
            {
              "height": 48,
            },
          ],
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#000",
            },
            [
              {
                "position": "absolute",
              },
              {
                "left": 0,
              },
              {
                "right": 0,
              },
              {
                "flexBasis": "0%",
                "flexGrow": 1,
                "flexShrink": 1,
              },
              {
                "textAlign": "center",
              },
              {
                "fontSize": 16,
                "lineHeight": 24,
              },
            ],
          ]
        }
      >
        Contact Detail
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#fff",
            "marginLeft": -8,
            "opacity": 1,
            "paddingBottom": 8,
            "paddingLeft": 8,
            "paddingRight": 8,
            "paddingTop": 8,
          }
        }
      >
        <RNSVGSvgView
          bbHeight={24}
          bbWidth={24}
          color="black"
          fill="none"
          focusable={false}
          height={24}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "flex": 0,
                "height": 24,
                "width": 24,
              },
            ]
          }
          tintColor="black"
          width={24}
        >
          <RNSVGGroup
            fill={null}
            propList={
              [
                "fill",
              ]
            }
          >
            <RNSVGPath
              d="M2 12h20M2 12l6-6M2 12l6 6"
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              propList={
                [
                  "stroke",
                  "strokeWidth",
                  "strokeLinecap",
                ]
              }
              stroke={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              strokeLinecap={1}
              strokeWidth={2}
            />
          </RNSVGGroup>
        </RNSVGSvgView>
      </View>
    </View>
  </View>
  <RCTScrollView
    keyboardShouldPersistTaps="handled"
    style={
      [
        {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
  >
    <View>
      <View
        style={
          [
            {
              "overflow": "hidden",
            },
            [
              {
                "width": 96,
              },
              {
                "height": 96,
              },
              {
                "marginRight": 16,
              },
              {
                "borderBottomLeftRadius": 9999,
                "borderBottomRightRadius": 9999,
                "borderTopLeftRadius": 9999,
                "borderTopRightRadius": 9999,
              },
              {
                "backgroundColor": "#d1d5db",
              },
              {
                "alignSelf": "center",
              },
              {
                "marginTop": 16,
              },
            ],
          ]
        }
      >
        <FastImageView
          defaultSource={null}
          resizeMode="cover"
          style={
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
      </View>
      <Text
        style={
          [
            {
              "color": "#000",
            },
            [
              {
                "paddingLeft": 16,
                "paddingRight": 16,
              },
              {
                "marginTop": 16,
              },
              {
                "fontWeight": "700",
              },
            ],
          ]
        }
      >
        First Name:
      </Text>
      <TextInput
        onChangeText={[Function]}
        placeholder="Enter First Name"
        style={
          [
            {
              "borderBottomWidth": 1,
              "borderLeftWidth": 1,
              "borderRightWidth": 1,
              "borderTopWidth": 1,
            },
            {
              "borderBottomColor": "#d1d5db",
              "borderLeftColor": "#d1d5db",
              "borderRightColor": "#d1d5db",
              "borderTopColor": "#d1d5db",
            },
            {
              "paddingLeft": 16,
              "paddingRight": 16,
            },
            {
              "marginLeft": 16,
              "marginRight": 16,
            },
            {
              "borderBottomLeftRadius": 6,
              "borderBottomRightRadius": 6,
              "borderTopLeftRadius": 6,
              "borderTopRightRadius": 6,
            },
            {
              "color": "#000",
            },
          ]
        }
        testID="TextInputFirstName"
        value="John"
      />
      <Text
        style={
          [
            {
              "color": "#000",
            },
            [
              {
                "paddingLeft": 16,
                "paddingRight": 16,
              },
              {
                "marginTop": 16,
              },
              {
                "fontWeight": "700",
              },
            ],
          ]
        }
      >
        Last Name:
      </Text>
      <TextInput
        onChangeText={[Function]}
        placeholder="Enter Last Name"
        style={
          [
            {
              "borderBottomWidth": 1,
              "borderLeftWidth": 1,
              "borderRightWidth": 1,
              "borderTopWidth": 1,
            },
            {
              "borderBottomColor": "#d1d5db",
              "borderLeftColor": "#d1d5db",
              "borderRightColor": "#d1d5db",
              "borderTopColor": "#d1d5db",
            },
            {
              "paddingLeft": 16,
              "paddingRight": 16,
            },
            {
              "marginLeft": 16,
              "marginRight": 16,
            },
            {
              "borderBottomLeftRadius": 6,
              "borderBottomRightRadius": 6,
              "borderTopLeftRadius": 6,
              "borderTopRightRadius": 6,
            },
            {
              "color": "#000",
            },
          ]
        }
        testID="TextInputLastName"
        value="Doe"
      />
      <Text
        style={
          [
            {
              "color": "#000",
            },
            [
              {
                "paddingLeft": 16,
                "paddingRight": 16,
              },
              {
                "marginTop": 16,
              },
              {
                "fontWeight": "700",
              },
            ],
          ]
        }
      >
        Age:
      </Text>
      <TextInput
        keyboardType="numeric"
        onChangeText={[Function]}
        placeholder="Enter Age"
        style={
          [
            {
              "borderBottomWidth": 1,
              "borderLeftWidth": 1,
              "borderRightWidth": 1,
              "borderTopWidth": 1,
            },
            {
              "borderBottomColor": "#d1d5db",
              "borderLeftColor": "#d1d5db",
              "borderRightColor": "#d1d5db",
              "borderTopColor": "#d1d5db",
            },
            {
              "paddingLeft": 16,
              "paddingRight": 16,
            },
            {
              "marginLeft": 16,
              "marginRight": 16,
            },
            {
              "borderBottomLeftRadius": 6,
              "borderBottomRightRadius": 6,
              "borderTopLeftRadius": 6,
              "borderTopRightRadius": 6,
            },
            {
              "color": "#000",
            },
          ]
        }
        testID="TextInputAge"
        value="20"
      />
    </View>
  </RCTScrollView>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#0f766e",
        "borderBottomLeftRadius": 6,
        "borderBottomRightRadius": 6,
        "borderTopLeftRadius": 6,
        "borderTopRightRadius": 6,
        "height": 48,
        "justifyContent": "center",
        "marginBottom": 16,
        "marginLeft": 16,
        "marginRight": 16,
        "marginTop": 16,
        "opacity": 1,
      }
    }
    testID="ButtonSave"
  >
    <View
      style={
        [
          {
            "backgroundColor": "#fff",
          },
          [
            {
              "backgroundColor": "transparent",
            },
          ],
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#000",
            },
            [
              {
                "color": "#fff",
              },
              {
                "textAlign": "center",
              },
              {
                "fontWeight": "700",
              },
              {
                "fontSize": 18,
                "lineHeight": 28,
              },
            ],
          ]
        }
      >
        Save
      </Text>
    </View>
  </View>
</View>
`;
